doc/sphinx/proof-engine/ltac.rst

```diff
@@ -1146,7 +1143,8 @@ Pattern matching on terms: match
       metavariables in the form :n:`?@ident`.  :g:`_` can be used to match
       irrelevant terms.  :ref:`Example<match_with_holes_ex>`.
 
-      .. todo PR Didn't understand the following paragraph well enough to revise
+      .. todo Didn't understand the following 2 paragraphs well enough to revise
+         would like to get an example for each
```

Comment:
Example from https://github.com/coq/coq/pull/12103#discussion_r435577241 :
Here is an example (perhaps not the best one, because it relies on the trick of `match a with b => c end` reducing the `match` away):
```coq
Ltac uncurry f :=
  lazymatch f with
  | fun (a : ?A) (b : ?B) => ?f (* note the lack of `@?` *)
    => let ab := fresh in
       uncurry (fun ab : A * B => match fst ab, snd ab with a, b => f end)
  | ?f => f
  end.

Goal True.
  let x := uncurry (fun (a b c : unit) (d e : nat) => (a, b, c, d, e)) in
  pose x.
  (* 1 subgoal (ID 38)

  p := fun H0 : unit * unit * unit * nat * nat =>
       (fst (fst (fst (fst H0))), snd (fst (fst (fst H0))),
       snd (fst (fst H0)), snd (fst H0), snd H0)
   : unit * unit * unit * nat * nat -> unit * unit * unit * nat * nat
  ============================
  True
   *)
```
Here is a variant that doesn't use that
```coq
Ltac uncurry f :=
  lazymatch (eval cbv beta in f) with
  | fun (a : ?A) (b : ?B) => ?f (* note the lack of `@?` *)
    => uncurry ((fun F (ab : A * B) => F (fst ab) (snd ab)) (fun (a : A) (b : B) => f))
  | ?f => f
  end.

Goal True.
  let x := uncurry (fun (a b c : unit) (d e : nat) => (a, b, c, d, e)) in
  pose x.
  (* 1 subgoal (ID 38)

  p := fun H0 : unit * unit * unit * nat * nat =>
       (fst (fst (fst (fst H0))), snd (fst (fst (fst H0))),
       snd (fst (fst H0)), snd (fst H0), snd H0)
   : unit * unit * unit * nat * nat -> unit * unit * unit * nat * nat
  ============================
  True
   *)
```
Does this explain it?

Note also that you are already making use of this feature in `_match_with_holes_ex:`

As for the second paragraph, here's the same tactic but written using second-order pattern matching:
```coq
Ltac uncurry f :=
  lazymatch (eval cbv beta in f) with
  | fun (a : ?A) (b : ?B) => @?f a b
    => uncurry (fun ab : A * B => f (fst ab) (snd ab))
  | ?f => f
  end.
```

Perhaps a better example of the former is swapping args:
```coq
Ltac swap_args f :=
  lazymatch f with
  | fun (a : ?A) (b : ?B) => ?f
    => constr:(fun (b : B) (a : A) => f)
  end.
```
But you can also do with with a neat trick of `@?` patterns:
```coq
Ltac swap_args f :=
  lazymatch f with
  | fun a b => @?f b a (* note the reversed order *)
    => f
  end.

Goal True.
  let f := swap_args (fun x y => x + y) in
  pose f. (* fun y x : nat => x + y : nat -> nat -> nat *)
```

Reply:
I think I'll leave this as a "todo" for future PRs.  The example is a bit long; I'd want to make it as simple as possible (but no simpler).

Reply:
Sounds good.  Perhaps link to this discussion in the todo, so that we can find it again (otherwise it'll be quite hard to find, I think).

Reply:
Yup, that's what I've been doing.  There are 8 such links so far.